[77] Improve test suite for RSAEncrypter/Decrypter

gigi-mohemian opened the pull request:
Resolves part of issue #77 by extending the RSA encrypter/decrypter test suite.

The test keypair is extended into a set of two keypairs, one for `Alice` and one for `Bob`. Additional tests are added to test encryption and decryption of payloads with these keys. Some tests are added for specific error cases, i.e. trying to decrypt with the wrong key.

Looking forward to your comments! 💻 👀

carol-mohemian commented on Tests/RSADecrypterTests.swift:
```diff
@@ -25,7 +25,11 @@ import XCTest
-    let cipherTextBase64URL = "YkzUN55RBG1igsrL-7ofPPWruTjxNMS3Bl1_a3i-dVKBjiN6kH88j8G3iB3eLWFxbiKZvx0LmkP7J65-frpOgF41SlltjJ62LEZOICm4q21Q4MNqL_Vf7kIjh8DziEJkElEz5W4flhI6YQ-9wW_PQ-coBIPIZiFlw6peKAolz8xcevbUmnqIH6A3hOFLK23J2cWDSWgxHEBIYtZ6whQCJYL4vq5lAFNaEoDaE_cgL6LItY4t-vR1exTJSOlCGAv4uM1Kelk6uitaFk2c0h79u3UpFN_wa02m_PPdgTguRdxwRsCpsQhOKmEakl8LR6NTbIrdB13UoL2tdybltVeUCw"
+    let cipherTextWithAliceKeyBase64URL = "RHLdaxC52ve0icBLGh9Xn_jNoV3uJlEuLRer8_7OOqY2QCGdWPb8OiuxxjEep8okfvlrw_eheEll6v8xdjpzP2zopVEBoNZ9J8BDKL1Y-tTi176jlGXBXKJdRitsp14sVHx7x-dW9FjiOHEn_CSm3bM2THAJOEtjjeF6vRvxpGxm1KruTbtZ37VTl-vM-e4STda5dCanRPKFemDPYQvjUAF_wuE-P-UeE0fpxDceuCRN3C-H8LL9TIrSETRV_CzPVf6Ki9zJpaZQAnqZh5ix0KhaHFMawV7TJetcKPbCzEHbxAF5ib16mxKFc5Th3QFS3eRKYxjaEdciXWVMCrGfXQ"
+    let cipherTextWithBobKeyBase64URL = "TwSS87gjm4cbFbda5klqW_Io2oCoo2zAxWJaU6L1jAeGEoL6u1Gw7vPdEzGGNvUliqsdBBCe8yQhnft9nqIZ9CD7enfVz6tFWvEJuVZbhy0eZad_LLFBFm8eDEgmlw7n7nAZxeww0WloffMmK6GPUE5SjQSWjCRAYA9xwacbxM8shH6hIwpelwV3f-QJr6M_hoAVkiqZJ2tFfiEciPhVVuqNZxxudMxaNj-bWfSipwh2T11kxYQD3EdVR4MDkyxz2JTUQc20OIf_F4WwQXJlbVNLTs4miDKHmw7Ddd45D2DJs_nH-bMQK8b6iH3nke15arP23rozuagF4SHhenApPw"
+    let cipherTextWithAliceKeyEmptyStringBase64URL = "hyOA/tmOclFkj31UPrRb1EnaRMhR5VZg5TrUyfLMtCUlh3grAva0+sSjqt6zSlWK06A6zUieV69aLRbJ0ZactTTqX2CFrhiZ5nUXhzuUya83VKBI0xrGkpQ8u1y2Iqgb+gbWsFJdJ41cSpZXRpc16Hhd3klTp7YydYZQUG//PLM5bn359kqpT8meJdGqTceehVxmdqTpVwukh/uqLOE8CBCrT7D/2t18mzApGpm/Su4bVbZggJ5g9MRPSnwgq1GjKNcMa1PKd+/OWB/rIeDmorT8dLrusGeLbwFCj1HEz4z5izamiBiyPh96G0m4ZhPtVhR4Fo3ARj9C037GroDQ2w=="
```
Is this base64URL? 🤔 I guess the padding(`==`) and the `/` isn't 😁

daniel-mohemian replied:
Nope, you're right. It's not base64url. @gigi-mohemian you can quickly convert it online [here](https://cryptii.com) or by hand replace all `/` with `_`, all `+` with `-` and remove the trailing `==`.

Or you just remove the `URL` from the variable name. 😄

daniel-mohemian replied:
Are the cipher texts are generated by openssl?

I know the original string was a nastly long thing as well, but now that we started touching the tests again, what do you think of staying below 120 characters per line (if possible). With strings its kind of easy to use [Swift's multiline string literals](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/StringsAndCharacters.html):

e.g.:

``` swift
let cipherText = """
    asdf\
    jklö
"""

daniel-mohemian commented on .gitignore:
```diff
@@ -101,3 +101,4 @@ fastlane/test_output
 
 .idea/
 sonar-reports/
+/JOSESwift.xcworkspace/xcshareddata
```
🙏

daniel-mohemian commented on Tests/RSADecrypterTests.swift:
```diff
@@ -35,28 +39,81 @@ class RSADecrypterTests: CryptoTestCase {
         XCTAssertEqual(plainText, message.data(using: .utf8))
     }
 
+    func testDecryptingAliceSecretWithBobKey() {
```
👌

gigi-mohemian commented:
@carol-mohemian @daniel-mohemian I addressed the issues mentioned. Please have a look again 👍

carol-mohemian commented on Tests/RSADecrypterTests.swift:
```diff
@@ -25,9 +25,24 @@ import XCTest
+-H8LL9TIrSETRV_CzPVf6Ki9zJpaZQAnqZh5ix0KhaHFMawV7TJetcKPbCzEHbxAF5ib16mxKFc5Th3QFS3eRKYxjaEdciXWVMCrGfXQ
+"""
+
+    let cipherTextWithBobKeyBase64 = """
```
Hmm, now this is a little confusing, why is `cipherTextWithBobKeyBase64` and cipherTextWithAliceKeyBase64` base64URL and `cipherTextWithAliceKeyEmptyStringBase64` "only" base64?

gigi-mohemian replied:
Ah damn, you're right. I'll clean this up asap.

gigi-mohemian replied:
It's base64 encoded now.

```
printf "The true sign of intelligence is not knowledge but imagination." | openssl rsautl -encrypt -pubin -inkey alice.pub.pem -out >(base64)
```

gigi-mohemian commented:
Sorry this took so long @daniel-mohemian @carol-mohemian ⏳ 

The tests for RSA encryption/decryption should be fine now, I re-created the cipher texts using Alice's and Bob's keys from scratch using `openssl`, since it's been so long 😅I wanted to make sure that everything is correct.

I'd like to do additional tests for AES and for the signer/verifier, but I think it's best to do it in another PR.

Please have another look and let me know what you think 😘

gigi-mohemian commented:
I agree as well! 😄 

Double-checked the RFC and updated the tests. Added the following:

*Encryption:*

- test if encryption of maximum length message `mLen` works
- test if encryption fo `mLen + 1` fails

*Decryption:*

- test if decryption of `len = k` works (i.e., does not throw a "length" error)
- test if decryption of `len = k-1` fails
- test if decryption of `len = k+1` fails

carol-mohemian commented on Tests/RSADecrypterTests.swift:
```diff
@@ -131,4 +131,56 @@ sVyhqpFuZQ6hhklG9lJr6OBBuk/+pcJYdHuYEuLnJhPeKqF/9xgMOU0e0xLMtkQW+IfDMlm0oAVavHrx
+        XCTAssertThrowsError(try decrypter.decrypt(testMessage)) { (error: Error) in
+            // Should throw "decryption failed", but
+            // should _not_ throw cipherTextLenghtNotSatisfied
+            XCTAssertNotEqual(error as? RSAError, RSAError.cipherTextLenghtNotSatisfied)
```
This test will fail as soon as #96 is done 🤔

daniel-mohemian replied:
Good catch!

carol-mohemian commented on Tests/RSADecrypterTests.swift:
```diff
@@ -131,4 +131,56 @@ sVyhqpFuZQ6hhklG9lJr6OBBuk/+pcJYdHuYEuLnJhPeKqF/9xgMOU0e0xLMtkQW+IfDMlm0oAVavHrx
+
+        let decrypter = RSADecrypter(algorithm: .RSA1_5, privateKey: privateKeyAlice2048!)
+        XCTAssertThrowsError(try decrypter.decrypt(testMessage)) { (error: Error) in
+            XCTAssertEqual(error as? RSAError, RSAError.cipherTextLenghtNotSatisfied)
```
Same #96 (just to give a heads up)
